pkg/controller/permissions/permissions_check.go

```diff
@@ -3,74 +3,134 @@ package permissions
 import (
 	"errors"
 	"fmt"
+	"net/http"
 	"strings"
 
 	"github.com/acorn-io/baaah/pkg/router"
 	"github.com/acorn-io/baaah/pkg/uncached"
 	apiv1 "github.com/acorn-io/runtime/pkg/apis/api.acorn.io/v1"
 	v1 "github.com/acorn-io/runtime/pkg/apis/internal.acorn.io/v1"
+	"github.com/acorn-io/runtime/pkg/config"
+	"github.com/acorn-io/runtime/pkg/imagerules"
+	"github.com/acorn-io/runtime/pkg/labels"
+	"github.com/acorn-io/runtime/pkg/profiles"
 	"github.com/acorn-io/runtime/pkg/tags"
+	"github.com/acorn-io/z"
 	"github.com/google/go-containerregistry/pkg/name"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+// CopyPromoteStagedAppImage copies the staged app image to the app image if
+// - the staged app image is set
+// - the permissions have been checked
+// - there are no missing permissions
+// - there are no image permissions denied (if ImageRoleAuthorizations are enabled)
+// - the image is allowed by the image allow rules (if enabled)
 func CopyPromoteStagedAppImage(req router.Request, resp router.Response) error {
 	app := req.Object.(*v1.AppInstance)
 	if app.Status.Staged.AppImage.ID != "" &&
 		app.Status.Staged.PermissionsChecked &&
-		len(app.Status.Staged.PermissionsMissing) == 0 {
+		len(app.Status.Staged.PermissionsMissing) == 0 &&
+		len(app.Status.Staged.ImagePermissionsDenied) == 0 &&
+		z.Dereference[bool](app.Status.Staged.ImageAllowed) {
 		app.Status.AppImage = app.Status.Staged.AppImage
 	}
 	return nil
 }
 
-func CheckImagePermissions(req router.Request, resp router.Response) error {
-	app := req.Object.(*v1.AppInstance)
-	if app.Status.Staged.AppImage.ID == "" ||
-		app.Status.Staged.AppImage.Digest == app.Status.AppImage.Digest ||
-		app.Status.Staged.PermissionsObservedGeneration == app.Generation {
-		return nil
-	}
+// CheckPermissions checks various things related to permissions
+// a) if the image is allowed by the image allow rules (if enabled)
+// b) if the permissions requested by all images in the app are
+//
+//	b.1) granted by the user (as set in the app spec)
+//	b.2) authorized by the image role authorizations (if enabled)
+func CheckPermissions(transport http.RoundTripper) router.HandlerFunc {
+	return func(req router.Request, _ router.Response) error {
```

Comment:
Looks like this transport argument is unused.

Reply:
Oh yeah, removed and put back in place during my debugging sessions - will use it again.